### PR TITLE
adds posture checks

### PR DESCRIPTION
**Pull request by andrewpmartinez**
- new callback options on ziti_options
- new timer to send checks every 20s
- debug output for missing callbacks if needed
- posture check submit on new service connect

### CODE REVIEW

File: inc_internal/ziti_ctrl.h [hunk 1]
```diff
@@ -63,6 +69,23 @@ void ziti_ctrl_get_well_known_certs(ziti_controller *ctrl, void (*cb)(char *, zi
 void ziti_ctrl_enroll(ziti_controller *ctrl, const char *method, const char *token, const char *csr,
                       void (*cb)(ziti_enrollment_resp *, ziti_error *, void *), void *ctx);
 
+void ziti_ctrl_pr_post_domain(ziti_controller *ctrl, char *id, char *domain,
+                              void (*cb)(void *, ziti_error *, void *),
```
Comment (smilindave26) [inline inc_internal/ziti_ctrl.h:73]: a typedef would be nice here for the callback type (especially when using from Swift, makes syntax much cleaner)

Comment (smilindave26) [reply inc_internal/ziti_ctrl.h:73]: tho we won't access these methods directly from Swift, so it wount make any difference :)

File: inc_internal/ziti_ctrl.h [hunk 2]
```diff
@@ -27,6 +27,11 @@ limitations under the License.
 extern "C" {
 #endif
 
+const char* const PC_DOMAIN_TYPE;
```
Comment (smilindave26) [inline inc_internal/ziti_ctrl.h]: Build fails on MacOS with:

```FAILED: tests/CMakeFiles/all_tests.dir/test_jwt.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DBUILD_DATE=Thu-11/05/2020-17:12:51-EST -DSOURCE_PATH_SIZE=3 -DZITI_BRANCH=add.posture.checks -DZITI_BUILDNUM=local -DZITI_COMMIT=02954c0 -DZITI_VERSION=0.17.3.1 -I../inc_internal -I../library/../includes -I_deps/uv-mbed-src/include -I_deps/http_parser-src -I_deps/libuv-src/include -I_deps/uv_link-src/include -isystem _deps/libsodium-build/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fPIE   -std=gnu++11 -MD -MT tests/CMakeFiles/all_tests.dir/test_jwt.cpp.o -MF tests/CMakeFiles/all_tests.dir/test_jwt.cpp.o.d -o tests/CMakeFiles/all_tests.dir/test_jwt.cpp.o -c ../tests/test_jwt.cpp
In file included from ../tests/test_jwt.cpp:19:
In file included from ../inc_internal/zt_internal.h:28:
../inc_internal/ziti_ctrl.h:30:19: error: default initialization of an object of const type 'const char *const'
const char* const PC_DOMAIN_TYPE;```

Comment (smilindave26) [reply inc_internal/ziti_ctrl.h]: define all these with `extern` and build continues...

Comment (smilindave26) [reply inc_internal/ziti_ctrl.h]: But build ends up failing with:

```FAILED: tests/all_tests
: && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  tests/CMakeFiles/all_tests.dir/all_tests.cpp.o tests/CMakeFiles/all_tests.dir/test_ziti_model.cpp.o tests/CMakeFiles/all_tests.dir/ctrl_tests.cpp.o tests/CMakeFiles/all_tests.dir/test_jwt.cpp.o tests/CMakeFiles/all_tests.dir/model_tests.cpp.o tests/CMakeFiles/all_tests.dir/test_metrics.cpp.o  -o tests/all_tests -L/Users/dhart/repos/netfoundry/ziti-sdk-c/build-macosx/lib -Wl,-rpath,/Users/dhart/repos/netfoundry/ziti-sdk-c/build-macosx/lib  library/libziti.a  _deps/uv-mbed-build/libuv_mbed.a  _deps/libuv-build/libuv_a.a  -lpthread  _deps/mbedtls-build/library/libmbedtls.a  _deps/mbedtls-build/library/libmbedx509.a  _deps/mbedtls-build/library/libmbedcrypto.a  _deps/libsodium-build/lib/libsodium.a  -lm && :
ld: warning: directory not found for option '-L/Users/dhart/repos/netfoundry/ziti-sdk-c/build-macosx/lib'
duplicate symbol '_posture_checks' in:
    tests/CMakeFiles/all_tests.dir/ctrl_tests.cpp.o
    tests/CMakeFiles/all_tests.dir/test_jwt.cpp.o```

File: inc_internal/zt_internal.h [hunk 3]
```diff
@@ -141,6 +141,18 @@ struct ziti_conn {
...
+    bool is_running;
+    char *sha_512_hash;
+    char **signers;
+    int num_signers;
+};
+
+struct posture_checks {
+    uv_timer_t timer;
+    double interval;
+} posture_checks;
```
Comment (smilindave26) [inline inc_internal/zt_internal.h]: Change to
```
struct posture_checks {
    uv_timer_t timer;
    double interval;
};
```

To fix the Mac/iOS `duplicate symbol '_posture_checks'` error

**Review by smilindave26 — APPROVED** [review]

File: library/posture.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+void ziti_send_posture_data(struct ziti_ctx *ztx) {
+    ZITI_LOG(DEBUG, "starting to send posture data");
+    const char *service_id;
+
+    NEWP(domainInfo, struct query_info);
+    NEWP(osInfo, struct query_info);
+    NEWP(macInfo, struct query_info);
+    struct query_info *procInfo = NULL;
+
+    NEWP(processes, model_map);
```
Comment (ekoby) [inline library/posture.c:77]: any reason to allocate vs using a stack variable?

File: library/posture.c [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+        if (ztx->opts->pq_process_cb != NULL) {
+            MODEL_MAP_FOREACH(path, info, processes) {
+                ztx->opts->pq_process_cb(ztx, info->query->id, path, ziti_handle_process);
+                free(info);
+            }
+        } else {
+            ZITI_LOG(DEBUG, "%s cb not set requested for: service %s, policy: %s, check: %s", PC_PROCESS_TYPE,
+                     osInfo->service->name, osInfo->query_set->policy_id, osInfo->query->id);
+            MODEL_MAP_FOREACH(path, info, processes) {
```
Comment (ekoby) [inline library/posture.c:157]: use model_map_clear() -- otherwise keys a leaked

File: programs/ziti-prox-c/proxy.c [hunk 6]
```diff
@@ -415,6 +415,30 @@ const char *my_configs[] = {
         "all", NULL
 };
 
+
```
Comment (ekoby) [inline programs/ziti-prox-c/proxy.c:418]: It think this has to be :axe: or provided real implementations

Comment (andrewpmartinez) [reply programs/ziti-prox-c/proxy.c:418]: Will remove.

File: library/posture.c (hunk 5 above)
Comment (andrewpmartinez) [reply library/posture.c]: Ahhh now nice! Done.

File: library/posture.c (hunk 4 above)
Comment (andrewpmartinez) [reply library/posture.c]: Fixed.

File: programs/ziti-prox-c/proxy.c (hunk 6 above)
Comment (andrewpmartinez) [reply programs/ziti-prox-c/proxy.c]: Removed.

**Review by ekoby — APPROVED** [review]

**Pull request merged**